codice/ddf
# [2.26.x] DDF-4729 adds WFS 1.1.0 sorting

#### What does this PR do?
Ports & refactors the WFS 2.0 sorting functionality into the WFS 1.1.0 source. Adds flag to enable/disable sorting (default to disable).  Adds mapping of DDF sort attributes to WFS attributes.


#### Who is reviewing it? 
@millerw8 
@jrnorth 
@glenhein 
@derekwilhelm 

#### Select relevant component teams: 
@codice/data 

#### Ask 2 committers to review/merge the PR and tag them here.
@jrnorth 
@emmberk 

#### How should this be tested?
1) build / install DDF
2) build DDF UI master (update ddf.version to 2.26.1-SNAPSHOT)
3) install DDF UI features into DDF per the README https://github.com/codice/ddf-ui/blob/master/README.md
4) Verify Intrigue UI is accessible
5) start up GeoServer
6) configure a Metacard to WFS Feature Map 
```
Feature type = {http://www.openplans.org/topp}states
Metacard title to feature property mapping = STATE_NAME
```
Save the config
7) configure a new WFS v1.1.0 Federated Source
```
WFS URL = http://<host>:<port>/geoserver/ows
Forced Feature Type = states
disable sorting true (should be default)
```
Save the config
8) From Intrigue, execute an anyText contains * search with sorting by “title” A-Z, source = WFS.  Verify the WFS results are returned.  Note, the results will not be sorted (as expected)
9) Update the WFS 1.1.0 Federated Source configuration and ENABLE sorting
10) repeat the same query.  The results should now be sorted by state name (Alabama, Arizona, etc).
11) verify paging works as expected and the results are still sorted
12) change the query to sort by “created” timestamp
13) Verify no results are returned (expected) and the heartbeat indicates a warning.  Tooltip will say “Something went wrong searching this source”.

#### Any background context you want to provide?

#### What are the relevant tickets?
Fixes: #4729 

#### Screenshots


#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
